File: java/1927/Solution.java
class Solution {
    public boolean sumGame(String num) {
        int blankCount = (int) num.chars()
            .filter(ch -> ch == '?')
            .count();

        if (blankCount % 2 != 0) {
            return true;
        }

        String[] nums = splitNum(num);

        String leftNum = nums[0];
        String rightNum = nums[1];

        return isNotMatchable(leftNum, rightNum);
    }

    private String[] splitNum(String num) {
        int chunkSize = num.length() / 2;
        return num.split("(?<=\\G.{" + chunkSize + "})");
    }

    private boolean isNotMatchable(String num1, String num2) {
        return digitOnlySum(replaceTwoBlankWithNine(num1)) != digitOnlySum(replaceTwoBlankWithNine(num2));
    }

    private int digitOnlySum(String num) {
        return num.chars()
            .filter(Character::isDigit)
            .map(ascii -> ascii - '0')
            .sum();
    }

    private String replaceTwoBlankWithNine(String num) {
        return num.replaceAll("\\?(\\d*)\\?", "9$1");
    }
}

File: python3/1927/Solution.py
import re

class Solution:
    def sumGame(self, num: str) -> bool:
        blank_count = num.count('?')

        if blank_count % 2 != 0:
            return True

        leftNum = num[:len(num)//2]
        rightNum = num[len(num)//2:]

        return self.isNotMatchable(leftNum, rightNum);

    def isNotMatchable(self, num1: str, num2: str) -> bool:
        return self.digitOnlySum(self.replaceTwoBlankWithNine(num1)) != self.digitOnlySum(self.replaceTwoBlankWithNine(num2));

    def digitOnlySum(self, num: str) -> int:
        digit_sum = 0

        for char in num:
            if char.isdigit():
                digit_sum += int(char)

        return digit_sum

    def replaceTwoBlankWithNine(self, num: str) -> str:
        return re.sub(r"\?(\d*)\?", r"9\1", num)
